private/lib/coordinates.bzl

```diff
@@ -1,103 +1,83 @@
 
     Assumes `coords` is in one of the following syntaxes:
-     * groupId:artifactId[:type[:scope]]:version
-     * groupId:artifactId[:version][@classifier][:type]
+     * group:artifact[:packaging[:classifier]]:version
+     * group:artifact[:version][:classifier][@packaging]
     """
     if not coords:
         return None
 
-    parts = coords.split(":")
-    nparts = len(parts)
-
-    if nparts < 2:
-        fail("Unparsed: %s" % coords)
-
-    # Both formats look the same for just `group:artifact`
-    if nparts == 2:
-        return struct(
-            groupId = parts[0],
-            artifactId = parts[1],
-            type = None,
-            scope = None,
-            version = None,
-            classifier = None,
-        )
-
-    # From here, we can be sure we have at least three `parts`
-    if _is_version_number(parts[2]):
-        return _unpack_gradle_format(coords)
-
-    return _unpack_rje_format(coords, parts)
+    pieces = coords.split(":")
+    group = pieces[0]
+    artifact = pieces[1]
+
+    if len(pieces) == 2:
+        return struct(group = group, artifact = artifact, version = "", packaging = None, classifier = None)
+
+    # Unambiguously the original format
+    if len(pieces) == 5:
+        packaging = pieces[2]
+        classifier = pieces[3]
+        version = pieces[4]
+        return struct(group = group, artifact = artifact, packaging = packaging, classifier = classifier, version = version)
+
+    # If we're using BOMs, the version is optional. That means at this point
+    # we could be dealing with g:a:p or g:a:v
+    is_gradle = pieces[2][0].isdigit()
+
+    if len(pieces) == 3:
+        if is_gradle:
+            if "@" in pieces[2]:
+                (version, packaging) = pieces[2].split("@", 2)
+                return struct(group = group, artifact = artifact, packaging = packaging, version = version, classifier = None)
+            version = pieces[2]
+            return struct(group = group, artifact = artifact, version = version, packaging = None, classifier = None)
+        else:
+            packaging = pieces[2]
+            return struct(group = group, artifact = artifact, packaging = packaging, version = "", classifier = None)
+
+    if len(pieces) == 4:
+        if is_gradle:
+            version = pieces[2]
+            if "@" in pieces[3]:
+                (classifier, packaging) = pieces[3].split("@", 2)
+                return struct(group = group, artifact = artifact, packaging = packaging, classifier = classifier, version = version)
+            classifier = pieces[3]
+            return struct(group = group, artifact = artifact, classifier = classifier, version = version, packaging = None)
+        else:
+            packaging = pieces[2]
+            version = pieces[3]
+            return struct(group = group, artifact = artifact, packaging = packaging, version = version, classifier = None)
+
+    fail("Could not parse maven coordinate: %s" % coords)
```

Comment:
Consider emitting the expected input format in this error message.

Reply:
Done.